### PR TITLE
Feat/refactor contract constraint expression

### PR DESCRIPTION
**Pull request by rbioteau**
Big PR sorry...
What it's in it ?
- New refactor abstraction for script refactoring : ScriptContainer
ScriptContainer is responsible to compute refactored script, update/remove script dependencies
ScriptContainer  is specialized for: 
*GroovyScript Expression
*Pattern Expression
*Condition Expression
*Constraint Expression

- Aside this major design change I have also change some weird algorithm (text replacement), reusin FindReplace action form jface.text

- There is also an improvement (I should have make another PR for that... sorry again) on Cell label provider to be notified on model change to support undo/redo properly.

Please merge https://github.com/bonitasoft/bonita-studio-sp/pull/6 in the same time

### CODE REVIEW

File: bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/jface/databinding/CustomTextEMFObservableValueEditingSupport.java [hunk 1]
```diff
@@ -124,14 +127,25 @@ protected void updateTextEditorFeedback(final IStatus status) {
     protected void validationStatusChanged(final IStatus status) {
         updateTextEditorFeedback(status);
         messageManager.removeAllMessages();
-        if (!status.isOK()) {
+        if (!status.isOK() && getViewer().isCellEditorActive()) {
             messageManager.addMessage("", status.getMessage(), null, new StatusToMessageSeverity(status).toMessageSeverity());
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/jface/databinding/CustomTextEMFObservableValueEditingSupport.java]: Why using empty string as first argument?

File: bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/predicate/ExpressionPredicates.java [hunk 2]
```diff
@@ -43,4 +48,31 @@ public boolean apply(final Expression input) {
...
+    }
+
+    private static boolean isElementIsReferencedInScript(final Expression expr, final EObject element) {
+        if (!expr.getReferencedElements().isEmpty()) {
+            for (final EObject o : expr.getReferencedElements()) {
+                if (EcoreUtil.equals(element, o)) {
+                    return true;
+                }
+                if (element instanceof Element && o instanceof Element && ModelHelper.isSameElement(element, o)) {
+                    return true && !ModelHelper.isReferencedElementIsInExpression(expr);
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/predicate/ExpressionPredicates.java]: "true &&" might surely be removed

File: bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/predicate/ExpressionPredicates.java [hunk 3]
```diff
@@ -43,4 +48,31 @@ public boolean apply(final Expression input) {
...
+                } else {
+                    if (element instanceof Parameter && o instanceof Parameter && ((Parameter) element).getName().equals(((Parameter) o).getName())) {
+                        return true && !ModelHelper.isReferencedElementIsInExpression(expr);
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/predicate/ExpressionPredicates.java]: "true &&" might surely be removed

File: bundles/plugins/org.bonitasoft.studio.expression.editor/src/org/bonitasoft/studio/expression/editor/viewer/ExpressionViewer.java [hunk 4]
```diff
@@ -1282,10 +1282,15 @@ public void setRefactorOperationToExecuteWhenUpdatingContent(final AbstractRefac
...
+            operation.clearItemToRefactor();
             final Object oldValue = getInput();
             if (oldValue instanceof Element && !newValue.equals(((Element) getInput()).getName())
                     || oldValue instanceof SearchIndex && !newValue.equals(((SearchIndex) getInput()).getName().getName())) {
-                operation.addItemToRefactor(newValue, getInput());
+                final SearchIndex searchIndex = (SearchIndex) getInput();
+                final SearchIndex newSearchIndex = EcoreUtil.copy(searchIndex);
+                newSearchIndex.getName().setContent(newValue);
+                newSearchIndex.getName().setName(newValue);
+                operation.addItemToRefactor(newSearchIndex, searchIndex);
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.expression.editor/src/org/bonitasoft/studio/expression/editor/viewer/ExpressionViewer.java]: c'est un peu relou de devoir créer un nouveau SearchIndex au lieu que ce soit le refactor qui s'en charge, non?

File: bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/core/emf/ObservableValueWithRefactor.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (C) 2014 BonitaSoft S.A.
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/core/emf/ObservableValueWithRefactor.java]: 2015 and BonitaSoft

File: bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/CompareScript.java [hunk 6]
```diff
@@ -75,56 +63,31 @@ public String getType() {
...
 
-    public void addChild(Object child) {
+    public void addChild(final Object child) {
         children.add(child);
     }
 
     @Override
     public boolean isEditable() {
-        return true;
+        return false;
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/CompareScript.java]: Why no more editable?

File: bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (C) 2014 BonitaSoft S.A.
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java]: 2015 et BonitaSoft

File: bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import org.eclipse.compare.ITypedElement;
+import org.eclipse.compare.structuremergeviewer.IStructureComparator;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IAdaptable;
+import org.eclipse.emf.edit.ui.provider.AdapterFactoryLabelProvider;
+
+public class LeftCompareScript extends CompareScript implements IStreamContentAccessor, IStructureComparator, IEditableContent, ITypedElement, IAdaptable,
+        IContentChangeNotifier {
+
+    LeftCompareScript(final ScriptContainer<?> script, final AdapterFactoryLabelProvider labelProvider) {
```
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java]: Why a constructor with Default package?

File: bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/CompareScript.java (hunk 6 above)
Comment (rbioteau) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/CompareScript.java]: Why the original script was editable in the compare view ? It seems not a valid use as we do not update the model with "left changes"

Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/CompareScript.java]: ok

File: bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/predicate/ExpressionPredicates.java (hunk 2 above)
Comment (rbioteau) [inline bundles/plugins/org.bonitasoft.studio.common/src/org/bonitasoft/studio/common/predicate/ExpressionPredicates.java]: Ouch. I just move an existing peace of code, didn't pay attention to this.

File: bundles/plugins/org.bonitasoft.studio.expression.editor/src/org/bonitasoft/studio/expression/editor/viewer/ExpressionViewer.java (hunk 4 above)
Comment (rbioteau) [inline bundles/plugins/org.bonitasoft.studio.expression.editor/src/org/bonitasoft/studio/expression/editor/viewer/ExpressionViewer.java]: I stop the refactoring here but, it's definitively something to change. A simple ReferenceDiff should be enough. Moreover this has nothing to do in the ExpressionViewer
I'll make another PR later about that.

File: bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java (hunk 8 above)
Comment (rbioteau) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java]: It's an internal implementation for CompareEditor.
What do you suggest ? protected ? I don't master the difference between package and protected

Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java]: yes protected, the default one - package - won't allow subclass placed in another package to see the constructor

Comment (rbioteau) [inline bundles/plugins/org.bonitasoft.studio.refactoring/src/org/bonitasoft/studio/refactoring/ui/LeftCompareScript.java]: Ok

File: bundles/plugins/org.bonitasoft.studio.expression.editor/src/org/bonitasoft/studio/expression/editor/viewer/ExpressionViewer.java (hunk 4 above)
Comment (apupier) [inline bundles/plugins/org.bonitasoft.studio.expression.editor/src/org/bonitasoft/studio/expression/editor/viewer/ExpressionViewer.java]: ok for another PR

**Pull request merged**